File: src/com/OpenSerenity/webDriver/WebDriverBrowser.java
package com.OpenSerenity.webDriver;

import com.OpenSerenity.core.Browser;
import com.OpenSerenity.core.DriverFactory;
import com.OpenSerenity.core.LocalStorage;
import com.OpenSerenity.core.TestContext;
import com.OpenSerenity.elements.BaseElement;
import com.OpenSerenity.elements.NativeElement;
import com.OpenSerenity.elements.WdNativeElement;
import com.OpenSerenity.functionalInterfaces.Func;
import com.OpenSerenity.functionalInterfaces.WaitCondition;
import com.OpenSerenity.utils.Waiter;
import org.apache.commons.io.FileUtils;
import org.apache.commons.lang3.StringUtils;
import org.openqa.selenium.*;

import java.io.File;
import java.io.IOException;
import java.util.Calendar;
import java.util.Date;
import java.util.Set;

public class WebDriverBrowser implements Browser {
    WebDriver driver;
    LocalStorage localStorage;

    public WebDriverBrowser() {
    }

    @Override
    public <TElement extends BaseElement> TElement findElement(Class<TElement> clz,
                                                               final String locator,
                                                               final Func<Browser> selectFrame)
            throws Exception {
        Waiter.Default().waitFor(new WaitCondition() {
            @Override
            public boolean invoke() {
                return driver.findElements(By.cssSelector(locator)).size() > 0;
            }
        });

        NativeElement nativeElement = new WdNativeElement(new Func<WebElement>() {
            @Override
            public WebElement invoke() throws Exception {
                selectFrame.invoke();
                return findWdNativeElement(locator);
            }
        });

        TElement element = clz.newInstance();
        element.setNativeElement(nativeElement);
        element.setLocator(locator);
        element.init();
        return element;
    }

    private WebElement findWdNativeElement(final String locator) throws Exception {
        Waiter.Default().waitFor(new WaitCondition() {
            @Override
            public boolean invoke() {
                return driver.findElements(By.cssSelector(locator)).size() > 0;
            }
        });
        return driver.findElement(By.cssSelector(locator));
    }

    @Override
    public void start() throws Exception {
        driver = DriverFactory.createBrowser();
        localStorage = new WebDriverLocalStorage(driver);
        maximize();
    }

    @Override
    public void stop() {
        driver.quit();
    }

    @Override
    public void open(String url) {
        driver.navigate().to(TestContext.configuration.getBaseDomain() + url);
    }

    @Override
    public void close() {
        driver.close();
    }

    @Override
    public void resizeTo(int width, int height) {
        driver.manage().window().setSize(new Dimension(width, height));
    }

    @Override
    public void maximize() {
        driver.manage().window().maximize();
    }

    @Override
    public void back() {
        driver.navigate().back();
    }

    @Override
    public void refresh() {
        driver.navigate().refresh();
    }

    @Override
    public void selectFrame(String locator) {
        if(StringUtils.isNotBlank(locator)) {
            driver.switchTo().frame(locator);
        } else {
            driver.switchTo().defaultContent();
        }
    }

    @Override
    public void selectWindow(String handle) {
        driver.switchTo().window(handle);
    }

    @Override
    public void closeWindow(String handle) {
        driver.switchTo().window(handle).close();
    }

    @Override
    public String getCurrentWindowHandle() {
        return driver.getWindowHandle();
    }

    @Override
    public Set<String> getAllWindowsHandles() {
        return driver.getWindowHandles();
    }

    @Override
    public String getTitle() {
        return driver.getTitle();
    }

    @Override
    public String getLocation() {
        return driver.getCurrentUrl();
    }

    @Override
    public String getMessageBoxText() {
        Alert alert = driver.switchTo().alert();
        return alert.getText();
    }

    @Override
    public void acceptMessageBox() {
        Alert alert = driver.switchTo().alert();
        alert.accept();
    }

    @Override
    public void dismissMessageBox() {
        Alert alert = driver.switchTo().alert();
        alert.dismiss();
    }

    @Override
    public void captureEntirePageScreenShot(String filename) {
        if (driver instanceof TakesScreenshot) {
            File tempFile = ((TakesScreenshot)driver).getScreenshotAs(OutputType.FILE);
            try {
                FileUtils.copyFile(tempFile, new File("screenshots/" + new Date() + ".png"));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    @Override
    public void createCookie(String name, String value) {
        Calendar cal = Calendar.getInstance();
        cal.setTime(new Date());
        cal.add(Calendar.MONTH, 1); // add 1 month

        Cookie cookie = new Cookie(name, value, null, cal.getTime());
        driver.manage().addCookie(cookie);
    }

    @Override
    public String getCookie(String name) {
        Cookie cookie = driver.manage().getCookieNamed(name);
        if(cookie == null) {
            return null;
        }
        return cookie.getValue();
    }

    @Override
    public void deleteCookie(String name, String path) {
        driver.manage().deleteCookieNamed(name);
    }

    @Override
    public void acceptAnyAlert() {
        //throw new UnsupportedOperationException(); //To change body of implemented methods use File | Settings | File Templates.
    }

    @Override
    public LocalStorage getLocalStorage() {
        return localStorage;
    }
}

File: src/scala/com/OpenSerenity/webDriver/WebDriverBrowser.scala
package com.OpenSerenity.webDriver

import com.OpenSerenity.core.{LocalStorage, TestContext, DriverFactory, Browser}
import com.OpenSerenity.utils.Waiter
import org.openqa.selenium._
import com.OpenSerenity.elements.{BaseElement, WdNativeElement, NativeElement}
import org.apache.commons.lang3.StringUtils
import scala.Predef.String
import java.io.{IOException, File}
import org.apache.commons.io.FileUtils
import java.util.{Calendar, Date}
import scala.collection.JavaConverters._
import scala.collection.mutable

class WebDriverBrowser extends Browser {
  def findElement[TElement <: BaseElement[TElement]](locator: String, selectFrame: () => Browser)(implicit m: scala.reflect.Manifest[TElement]): TElement = {
    Waiter.default.waitFor(() => driver.findElements(By.cssSelector(locator)).size > 0)
    val nativeElement: NativeElement = new WdNativeElement(() => {
        selectFrame()
        findWdNativeElement(locator)
    })
    val element: TElement = m.runtimeClass.newInstance().asInstanceOf
    element.setNativeElement(nativeElement)
    element.setLocator(locator)
    element.init
    return element
  }

  private def findWdNativeElement(locator: String): WebElement = {
    Waiter.default.waitFor(() => driver.findElements(By.cssSelector(locator)).size > 0)
    return driver.findElement(By.cssSelector(locator))
  }

  def start {
    driver = DriverFactory.createBrowser
    localStorage = new WebDriverLocalStorage(driver)
    maximize
  }

  def stop {
    driver.quit
  }

  def open(url: String) {
    if (url.startsWith("http")) {
      driver.navigate.to(url)
      return
    }
    driver.navigate.to(TestContext.configuration.getBaseDomain + url)
  }

  def close {
    driver.close
  }

  def resizeTo(width: Int, height: Int) {
    driver.manage.window.setSize(new Dimension(width, height))
  }

  def maximize {
    driver.manage.window.maximize
  }

  def back {
    driver.navigate.back
  }

  def refresh {
    driver.navigate.refresh
  }

  def selectFrame(locator: String) {
    if (StringUtils.isNotBlank(locator)) {
      driver.switchTo.frame(locator)
    }
    else {
      driver.switchTo.defaultContent
    }
  }

  def selectWindow(handle: String) {
    driver.switchTo.window(handle)
  }

  def closeWindow(handle: String) {
    driver.switchTo.window(handle).close
  }

  def getCurrentWindowHandle: String = {
    return driver.getWindowHandle
  }

  def getAllWindowsHandles: mutable.Set[String] = driver.getWindowHandles().asScala

  def getTitle: String = {
    return driver.getTitle
  }

  def getLocation: String = {
    return driver.getCurrentUrl
  }

  def getMessageBoxText: String = {
    val alert: Alert = driver.switchTo.alert
    return alert.getText
  }

  def acceptMessageBox {
    val alert: Alert = driver.switchTo.alert
    alert.accept
  }

  def dismissMessageBox {
    val alert: Alert = driver.switchTo.alert
    alert.dismiss
  }

  def captureEntirePageScreenShot(filename: String) {
    if (driver.isInstanceOf[TakesScreenshot]) {
      val tempFile: File = (driver.asInstanceOf[TakesScreenshot]).getScreenshotAs(OutputType.FILE)
      try {
        FileUtils.copyFile(tempFile, new File("screenshots/" + new Date + ".png"))
      }
      catch {
        case e: IOException => {
          e.printStackTrace
        }
      }
    }
  }

  def createCookie(name: String, value: String) {
    val cal: Calendar = Calendar.getInstance
    cal.setTime(new Date)
    cal.add(Calendar.MONTH, 1)
    val cookie: Cookie = new Cookie(name, value, null, cal.getTime)
    driver.manage.addCookie(cookie)
  }

  def getCookie(name: String): String = {
    val cookie: Cookie = driver.manage.getCookieNamed(name)
    if (cookie == null) {
      return null
    }
    return cookie.getValue
  }

  def deleteCookie(name: String, path: String) {
    driver.manage.deleteCookieNamed(name)
  }

  def acceptAnyAlert {
  }

  def getLocalStorage: LocalStorage = {
    return localStorage
  }

  private[webDriver] var driver: WebDriver = null
  private[webDriver] var localStorage: LocalStorage = null
}

